src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,247 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Composition;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.FindSymbols;
-using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.FindSymbols.SymbolTree;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
 using Microsoft.CodeAnalysis.SolutionCrawler;
 using Roslyn.Utilities;
+using static Roslyn.Utilities.PortableShim;
 
 namespace Microsoft.CodeAnalysis.IncrementalCaches
 {
-    [ExportIncrementalAnalyzerProvider(WorkspaceKind.Host), Shared]
-    internal class SymbolTreeInfoIncrementalAnalyzerProvider : IIncrementalAnalyzerProvider
+    [Shared]
+    [ExportIncrementalAnalyzerProvider(WorkspaceKind.Host)]
+    [ExportWorkspaceServiceFactory(typeof(ISymbolTreeInfoCacheService))]
+    internal class SymbolTreeInfoIncrementalAnalyzerProvider : IIncrementalAnalyzerProvider, IWorkspaceServiceFactory
     {
+        private class ProjectInfo
+        {
+            public readonly VersionStamp VersionStamp;
+            public readonly SymbolTreeInfo SymbolTreeInfo;
+
+            public ProjectInfo(VersionStamp versionStamp, SymbolTreeInfo info)
+            {
+                VersionStamp = versionStamp;
+                SymbolTreeInfo = info;
+            }
+        }
+
+        private class MetadataInfo
+        {
+            public readonly DateTime TimeStamp;
+            public readonly SymbolTreeInfo SymbolTreeInfo;
+            public readonly HashSet<ProjectId> ReferencingProjects;
+
+            public MetadataInfo(DateTime timeStamp, SymbolTreeInfo info, HashSet<ProjectId> referencingProjects)
+            {
+                TimeStamp = timeStamp;
+                SymbolTreeInfo = info;
+                ReferencingProjects = referencingProjects;
+            }
+        }
+
+        // Concurrent dictionaries so they can be read from the SymbolTreeInfoCacheService while 
+        // they are being populated/updated by the IncrementalAnalyzer.
+        private readonly ConcurrentDictionary<ProjectId, ProjectInfo> _projectToInfo = new ConcurrentDictionary<ProjectId, ProjectInfo>();
```

Comment:
These are singletons at the workspace level.  So i'm not concerned about a few extra object[] allocations.